### PR TITLE
Start a thrash test suite for the collection node

### PR DESCRIPTION
**Pull request by mikezucc**
Im creating a thrash test suite for the collection node to try and reproduce some very fickle, hard to reproduce issues. 

tl;dr:
- Split up existing table view thrash tests to have a common utility
- Created a collection view/node (UIKit reference flag) thrash test

What do yall think? I want to add a test that has multiple collection views in memory to test the way UIKit manages index paths globally.

### CODE REVIEW

**Comment by mikezucc** [comment]
The issue I was investigating: global resource of NSIndexPaths falling out of sync with UICollectionViewDataSource mid collection view update causing a crash inside UIKit

I think this multithreading issue in UIKit is exacerbated when the main run loop gets busied, with a user typing on the keyboard and network interrupts triggering further collection view reloads. What are some strategies to simulate this through the tests?

**Comment by mikezucc** [comment]
I think the XCTest runner also may not give the most accurate, true replication of the aforementioned problems since the test runner manipulates the run loop on its own so I put in a flag to prevent usage of XCTest.waitWithExpectation on some test runs, but its not entirely avoidable for the dispatch main thrash test.

from the docs:
```
 * -waitForExpectationsWithTimeout:handler: runs the run loop while handling events until all expectations
 * are fulfilled or the timeout is reached. Clients should not manipulate the run
 * loop while using this API.
```

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/384e654d1ce0b3bf1e420d51d08815e8/raw/b3bc6fe6eeec5a4a93ecc29e1228842a5506d438/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Tests/ASCollectionViewThrashTests.mm [hunk 1]
```diff
@@ -0,0 +1,184 @@
+//
+//  ASTableViewThrashTests.mm
```
Comment (nguyenhuy) [inline Tests/ASCollectionViewThrashTests.mm]:
```suggestion
//  ASCollectionViewThrashTests.mm
```

File: Tests/ASCollectionViewThrashTests.mm [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+      ASThrashTestItem *item = data[i].items[j];
+      ASThrashTestNode *node = (ASThrashTestNode *)[collectionView nodeForItemAtIndexPath:indexPath];
+      XCTAssertEqualObjects(node.item, item, @"Wrong node at index path %@", indexPath);
+    }
+  }
+}
+
+#pragma mark Test Methods
+
+// Disabled temporarily due to issue where cell nodes are not marked invisible before deallocation.
```
Comment (nguyenhuy) [inline Tests/ASCollectionViewThrashTests.mm]: Nit: Remove the comment since the test is enabled now.

File: Tests/ASCollectionViewThrashTests.mm [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  NSURL *caseURL = [[NSBundle bundleForClass:[self class]] URLForResource:@"ASThrashTestRecordedCase" withExtension:nil subdirectory:@"TestResources"];
+  NSString *base64 = [NSString stringWithContentsOfURL:caseURL encoding:NSUTF8StringEncoding error:NULL];
+
+  _update = [ASThrashUpdate thrashUpdateWithBase64String:base64];
+  if (_update == nil) {
+    return;
+  }
+
+  ASThrashDataSource *ds = [[ASThrashDataSource alloc] initCollectionViewDataSourceWithData:_update.oldData];
+  // why is ds.collectionView.test_enableSuperUpdateCallLogging available on table view but now collection view>?
```
Comment (nguyenhuy) [inline Tests/ASCollectionViewThrashTests.mm]: We just didn't need it before. May worth adding it if you think it'd be helpful.

File: Tests/ASThrashUtility.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
+//
+//  ASTableViewThrashTests.mm
```
Comment (nguyenhuy) [inline Tests/ASThrashUtility.h]:
```suggestion
//  Tests/ASThrashUtility.h
```

File: Tests/ASCollectionViewThrashTests.mm [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+#import <AsyncDisplayKit/ASCollectionView.h>
+#import <stdatomic.h>
+
+#import "ASThrashUtility.h"
+
+@interface ASCollectionViewThrashTests : XCTestCase
+
+@end
+
+@implementation ASCollectionViewThrashTests  {
```
Comment (nguyenhuy) [inline Tests/ASCollectionViewThrashTests.mm]: Per [our coding convention](https://github.com/TextureGroup/Texture/blob/master/CONTRIBUTING.md), brackets should be on the following line (and ASTableViewThrashTests.mm doesn't adhere to it). Can you update the rest of this PR to follow this rule?

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/5e515bb3c57d379082b5b03b8f827a21/raw/bc9e0213f926990cdfa0014c577149b945c71a41/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Pull request merged**